casper/src/main/rholang/MakeMint.rho

```diff
@@ -47,18 +49,27 @@ in {
                 }
               }
             } |
-            contract thisPurse(@"deposit", @amount, @src, success) = {
-              new result in {
-                @src!(*decr, amount, *result) | //call src decr function.
-                for(@decrSuccess <- result) {
-                  if (decrSuccess) {
-                    balance!("add", amount, *success) // add transferred amount to this purse
+
+            contract thisPurse(@"deposit", @amount, @src, successCh) = {
+              new ch in {
+                balance!("add", amount, *ch) |
+                for(@addSuccess <- ch) {
+                  if(addSuccess) {
+                    @src!(*decr, amount, *ch) |
+                    for(@decrSuccess <- ch) {
+                      if (decrSuccess) {
+                        successCh!(true)
+                      } else {
+                        balance!("sub", amount, *successCh)
```

Comment:
So in case you fail, you subtract the amount that you added? Why not just never add the amount in the first place?

Reply:
The failure comes from the balance, which is a NonNegativeNumber, if I was to check before add I would duplicate the code in NonNegativeNumber